File: src/com/easy/HappyNumber_202.java
/**
 * 
 */
package com.easy;

import java.util.HashSet;
import java.util.Set;

/**
 * @author Chris Lin
 * @version 1.0
 */
public class HappyNumber_202 {
    // 2019年7月11日
    // Use HashSet to keep judgment
    // Time Complexity: O(n)
    // Space Complexity:O(1)
    // Runtime: 2 ms, faster than 60.39%
    // Memory Usage: 33.6 MB, less than 6.40%
    public boolean isHappy(int n) {
        if (n < 1) {
            return false;
        }

        Set<Integer> mSet = new HashSet<>();
        while (n != 1 && !mSet.contains(n)) {
            mSet.add(n);
            int mNewN = 0, mTmp = 0;
            while (n > 0) {
                mTmp = n % 10;
                n /= 10;
                mNewN += mTmp * mTmp;
            }
            
            n = mNewN;
        }
        return (n == 1);
    }
}

File: Golang_src/easy/happyNumber_202.go
package easy

// Use slow fast pointer:
// Time Complexity: O(logn)
// Space Complexity:O(1)
// Runtime: 3 ms, faster than 45.61%
// Memory Usage: 1.9 MB, less than 100.00%
func isHappy(n int) bool {
	helper := func(num int) int {
		sum := 0
		for num > 0 {
			tmp := num % 10
			sum += tmp * tmp
			num /= 10
		}

		return sum
	}

	slow, fast := n, helper(n)

	for fast != 1 && slow != fast {
		slow = helper(slow)
		fast = helper(helper(fast))
	}

	return fast == 1
}
